Create Multichain-proposal.md

rossixiluo opened the pull request:
feature: add multichain proposal markdown file

rossixiluo committed:
Create Multichain-proposal.md

feature: add multichain proposal markdown file

manugunther commented on Multichain-proposal.md:
```diff
@@ -0,0 +1,112 @@
+
+Multichain was born as Anyswap on the 20th July 2020 to service the clear needs of different and diverse blockchains to communicate with each other. Each blockchain has its own unique services that it provides, its own community and its own development ecosystem. For our industry to reach the next level for consumers, we need a fast, secure, inexpensive and reliable way to exchange value, data an ...
+
+The solutions developed by Multichain allow almost all blockchains to inter-operate. There is no restriction to Ethereum like chains (e.g. Binance Smart Chain), or different Layer 2 chains requiring finality to Ethereum (e.g. Polygon), or a network of Parachains (e.g. Moonbeam in the PolkaDot system), or Bitcoin types of chain (e.g. Litecoin), or COSMOS chains (e.g. Terra). These are either now a ...
```
Could you share some documentation/specification of Litecoin bridging? Probably it's useful to link it given the model similarities.

rossixiluo replied:
You can check out the BTC repo https://github.com/anyswap/CrossChain-Router/tree/feature/btc

manugunther commented on Multichain-proposal.md:
```diff
@@ -0,0 +1,112 @@
+
+- User sends **k** *wrapped E* to the SMPC wallet.
+- SMPC nodes send **k** *tokens E* to the user wallet on the ETH side.
+- k *wrapped tokens E will be burnt when there is incoming transaction.*
```
Do we need some mechanism to handle possible rollbacks? In particular on every scenario were we need to burn wrapped tokens.

rossixiluo replied:
Only after the transactions reach the secure block height, the Multichain protocol can proceed the bridging request.

manugunther commented on Multichain-proposal.md:
```diff
@@ -0,0 +1,112 @@
+
+Because the value of token in Cardano is limited to int 64, we need to consider the decimal convert in the 3 categories. 
+
+a, If the total supply can be expressed by int 64 with the default decimal. The wrapped tokens on Cardano will keep the default decimal as native token.
```
What do you mean by "default decimal"?

rossixiluo replied:
Default decimal here means the decimal of the token on the source chain.

manugunther replied:
Let's change it then, "default" is confusing.

manugunther commented on Multichain-proposal.md:
```diff
@@ -0,0 +1,112 @@
+
+Because the value of token in Cardano is limited to int 64, we need to consider the decimal convert in the 3 categories. 
+
+a, If the total supply can be expressed by int 64 with the default decimal. The wrapped tokens on Cardano will keep the default decimal as native token.
```
If with "default decimal" you meant that if the total supply with the maximum precision can be expressed in int64, then this condition would mean that you could transfer the total supply in a single transaction. Is that what you want?
Remember that the limit of int64 is for representing values for each *utxo*, so you have a limit of tokens per transaction. It's not a limit about the total supply in Cardano side.

rossixiluo replied:
Yes, we considered the extreme situation that the total supply in a single transaction.

manugunther commented on Multichain-proposal.md:
```diff
@@ -0,0 +1,112 @@
+
+2, Rate
+
+The Multichain protocol does not send two users in one transaction. All the transactions will be processed one by one. So there is only one transaction (from other chains to Cardano) in one block.
```
What about congestion? It will be a problem if you have multiple users trying to bridge in a short period.

rossixiluo replied:
In this scenario, multiple transactions need to queue. However according to our current observation, it should not be a persisting issue.

manugunther commented on Multichain-proposal.md:
```diff
@@ -0,0 +1,112 @@
+
+3, Minimum ADA for every UTXO
+
+In Cardano, every UTXO must contain a minimal amount of ADA. After evaluation, extra fee equivalent to 2 ADA will be added to the base fee for every transaction from other chains to Cardano. 
```
So you need some mechanism for updating the fees on Ethereum side depending on the ADA/ETH rate. How would you manage that?

rossixiluo replied:
In this stage, the fee is adjusted periodically offchain. In the future version, this should be managed on chain with oracle.

manugunther replied:
Let's add that to the document.

manugunther commented on Multichain-proposal.md:
```diff
@@ -0,0 +1,112 @@
+# Multichain Proposal for Cardano
```
Flow diagrams of each use case would be great to fully understand the proposal.

rossixiluo replied:
Please check this flow diagram with comments. Thanks
[Cardano Flow.pdf](https://github.com/input-output-hk/Bridges-working-group/files/9525098/Cardano.Flow.pdf)

manugunther replied:
It would be much better to have flow diagrams corresponding with each use case, showing all the actions needed on the Cardano side:

* Case 1: Send Cardano native token to Ethereum.
* Case 2: Send back a wrapped token from Ethereum to Cardano.
* Case 3: Send Ethereum token to Cardano.
* Case 4: Send back a wrapped token from Cardano to Ethereum.

For each case specify what is the process if it's the first time a token is bridged (it will imply defining a new minting policy)

rossixiluo committed:
Update Multichain-proposal.md

Change the word of "default" to avoid confusing

rossixiluo committed:
CIP Cardano Multichain Router

This is detailed CIP for Cardano Multichain Router. User scenarios, specification, low level design, code base guide and flow charts are included for convenient community review.

rossixiluo committed:
Merge pull request #10 from input-output-hk/main

CIP Cardano Multichain Router

rossixiluo committed:
add flow chat pdf

add flow chat pdf

rossixiluo committed:
flow charts

flow charts

rossixiluo committed:
Add files via upload

rossixiluo committed:
Delete CIP%20Cardano%20Multichain%20Router%20b02eb0e3a2bf408a85bdbe6a4ce5603b directory

rossixiluo committed:
flow chart

flow chart

rossixiluo committed:
Add files via upload

rossixiluo committed:
Delete cardano_-_eth.pdf

rossixiluo committed:
Delete cardano_-_eth 1.pdf

rossixiluo committed:
Delete How_offchain_cached_utxo_work.pdf

rossixiluo committed:
Delete Untitled.png

rossixiluo committed:
Update CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md

rossixiluo committed:
Delete cardano_-_eth 1.pdf

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+(2) Each participant node does not need to approve the request command, which is agreed by default.
+
+The SMPC network is responsible for signing transactions on each supported blockchain to perform a multitude of tasks comprising the management of asset accounts and smart contracts. A Threshold, will be needed to sign the transaction. This Threshold Signature (TSS) is denoted as (t, n), such as (9,15), (15,21), (21,31) etc., where t is the number of nodes needed to sign, and n is the total numbe ...
```
Just out of curiosity, I haven't yet read the full paper on threshold ECDS. What happens if the participants would change? E.g. extending from (9,15) -> (15,21). Wouldn't that change the pubkey signature key, which is used by the policy to mint the wrapped tokens?

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+# User Cases
+
+Tokens on different chains can be bridged to Cardano, but let’s simplify it to the scenario between Ethereum and Cardano.
```
I'm assuming this is already handled, just in case and it might just my ignorance :). Let's imagine a hypothetical scenario. We have 3 chains:
1. C - cardano
2. D - some other chain where multichain works
3. E - ethereum

Setup
* user A moved some USDC from E -> C
* user B moved some USDC from E -> D

Now user B would want to move USDC from D -> C .

I'm assuming both userA and user B should receive the same wUSDC tokens on cardano. Are these pairings somewhere stored? who manages them? can they be changed/unlinked similar to how (no)madUSDC was disconnected between ETH & Milkomeda?

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+3, After the verification, the SMPC nodes work together to sign a transfer transaction on Ethereum.
+
+4, The SMPC nodes send the native USDC to user’s address on Ethereum. 
```
There seems to be a step missing around burning the wUSDC on the cardano side. Or those will remain on some address for later use and "garbage collected" ?

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
````diff
@@ -0,0 +1,395 @@
+```json
+{
+    "metadata": {
+        "key": "123",
````
Leaving here the link to the metadata registry: https://github.com/cardano-foundation/CIPs/blob/0638a723343ecac5f4420f44f10053f8a631181a/CIP-0010/registry.json#L1

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
````diff
@@ -0,0 +1,395 @@
+```solidity
+1, Verify the safe block height of the transaction.
+2, Verify the metadata has dedicated key and value is bind：toChainId.
+3, Verify the token in the transaction is in the supported list.
````
out of curiosity where would this list be stored? Hardcoded public registry/smart contracts/offchain private data?

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
````diff
@@ -0,0 +1,395 @@
+5, Check the amount is in the legitimate range. 
+```
+
+### Building Transaction
````
It'd be good to specify what kind of transaction :) (minting, user sending funds, bridge sending funds, etc.) I don't think this applies to all

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
````diff
@@ -0,0 +1,395 @@
+5, Calculate the amount based on these information.
+
+```solidity
+Suppose the elements: value：11111111（the decimal on source chain is:6）  fee:111111000000000000(default decimal of fee is:18), the decimal of destination:18
+1. Change the decimal of fee from 18 to the decimal onn source chian，18->6
+	fee:111111000000000000->111111
````
a few typos here and there. It would be good to explicitely define. Assuming from the description
**fee token** - the token that's bridged. E.g. when moving USDC from cardano to ethereum. Some wUSDC will be retained by the MPCs? based on what is this fee calculated?

**transaction fee** - paid on the chains native asset (ETH, ADA). I'm assuming this won't be part of the protocol on the source chain, on the destination chain, especially on ETH it can make bridging quite expensive if I'm not mistaken. So what would happen, if the value that the user tries to bridge is lower than the fees? Will they receive their tokens back or lost forever?

I think getting the price of eth assets should already be doable and working. Getting cardano native asset's value much harder and might require some thought. Unless the tokens are listed on some central exchanges and since in general the liquidity is relatively low the price of the cardano native assets might be very unstable and hard to estimate from multiple DEXes.

Apart from this the user would also need to cover for **minAda** in both directions. Will leave some comments in that section.

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+	cardano-cli transaction build-raw \
+	--tx-in txHash#0 \
+	--tx-in ...\
+	--tx-out $(reveiverAddr)+1500000+"targetToken amount" \
```
will this 1.5 ADA be constant or will it depend on the token (the actual minada depends on the length of the token name). In the description below there is 2 ADA as extra fee to cover the minada requirements.

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+### Minting policy
+
+The minting policy in the Cardano integration is the same as that in other integration. 1:1 Pegging minting. The Multichain protocol only mint the wrapped tokens as much as the locked native tokens. 
```
if it's a simple script it would be good to have the definition here :)

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+### Minimum ADA for every UTXO
+
+On Cardano, every UTXO must contain a minimal amount of ADA. After evaluation, extra fee equivalent to 2 ADA will be added to the base fee for every transaction from other chains to Cardano.
+
+For example. User send 1000 USDC from Ethereum to Cardano. The base fee is 0.1%, which is 1 USDC. The extra fee is 0.9USDC, which is equivalent to 2 ADA. The total fee is 1.9USDC. The user will receive 998.1USDC and 2 ADA on Cardano. The extra fee will adjust according to the ADA price fluctuation. 
+
+To ensure there is enough ADA on SMPC address to send transaction to users, the ADA balance on SMPC address needed to be monitored by Gas Fee Monitor Mechanism, which is monitoring the gas fees on all the SMPC addresses sending asset or calling smart contracts on chains.
+
+Minimum bridging amount will be set, bridging amount below the transaction will be ignored by Multichain protocol.
```
I think 2 things should be clarified:
1. When bridging from Cardano -> Ethereum, it's the user who needs to provide the minAda. Will this be deducted from their fees or this will be an additional fee the user needs to count with
2. What happens when the transaction is ignored. The MPC address would still own those tokens and minAda that the user sent there. The user can't claim it back. The MPC might need to burn it (if wrapped asset) otherwise it won't be pegged to the Eth side.

TechRiderWR replied:
Partial reaction from Luoxi
>Usually, when users use the frontend to build the bridging transaction, it does not allow user to build transaction with amount below the minimum amount. If user insist to send the amount below the minimum to the MPC address, the asset is lost and locked on MPC.

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+### Concurrency
+
+ Multichain protocol use transactions chaining on Cardano to improve the concurrency. 
```
it'd be good to specify how exactly :)

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+If the balance is not enough, the SMPC nodes will mint the lacking amount.
+
+When users bridged wrapped token on Cardano from Cardano to Ethereum, the wrapped tokens will be locked on the MPC address. A independent thread periodically burn the tokens locked on MPC address every week. 
```
Returning to the point above with the ignored swaps. So these tokens will be burnt, so the MPCs will effectively begin to own the tokens locked on the ETH side. Is that correct?

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+### Configuraton
+
+All the configuration is stored on a onchian smart contract.
```
It'd be good to define the mapping for cardano as well. E.g. in the `TokenConfig` if it's also supposed to give you the cardano information, would that be the `policyId.assetName` ?

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+# Flow Charts
+
+[How offchain cached utxo work.pdf](CIP%20Cardano%20Multichain%20Router%20b02eb0e3a2bf408a85bdbe6a4ce5603b/How_offchain_cached_utxo_work.pdf)
```
I'm not sure I understand what this is trying to depict. If it's just for some caching of utxos offchain, I don't think it's necessary to include here. That's more of an implementation detail (also periodically the offchain utxos would need to be verified in that case). For the rest below going to assume this has to do something with the transaction chaining. In that case the step after `onchain utxo` saying `Filter out utxo that is not in offchain utxo` doesn't seem to be enough. You would also need to filter out all utxos that are already considered spent. For example you start with 3 utxos onchain:
A - 20 ADA
B - 30 ADA
C - 10 ADA

Somebody should receive 40 ADA, so the bridge will create a TX without fees:
A + B -> D (10 ada) + user compensation (40 ada)

This tx is submitted but did not yet appear on chain. So you would still see the above 3 utxos on-chain, and 1 in cache?
*D - 10 ADA

So practically you would have:
C - 10 ADA
*D - 10 ADA
utxos available to build new transactions, and you need to filter out A,B when creating new transactions.

TechRiderWR replied:
A similar query was already answered by Luoxi. It'd be good to incorporate into this document as well:

> > This is very unclear to me. I’m guessing it depends on the definition of what’s cached. The formulation seems to suggest that on-chain and cached would be exclusive sets? This might get quite complicated if tx chaining should be introduced. Some of the utxos on-chain could have already been spent in the already submitted chain of transactions. I think this part might need some more precise definition.

> Yes, it’s quite complicate. When the node stores the cached map utxo, it also stores the input that has been used. When new UTXO needed to retrieve onchain, all the used utxo will be excluded before the merge. If transactions failed, they will be retried with new UTXO.

Additionally rollbacks and retries will need to be accounted for.

TechRiderWR commented on Multichain-proposal.md:
```diff
@@ -0,0 +1,112 @@
+
+The SMPC nodes will manage a single address, which will store the cardano native tokens bridged to ethereum and will mint the wrapped tokens. The only scripts needed are related to the minting policy for those wrapped tokens. This minting policy would simply check the signature of the SMPC wallet.
+
+There will be no scripts involved on the Cardano side. The SMPC nodes will manage a single wallet, and sign the transactions. 
```
careful with wording `scripts` -> `plutus smart contracts`. Technically the multisig minting script is considered a native script.

TechRiderWR commented on Multichain-proposal.md:
```diff
@@ -0,0 +1,112 @@
+
+2, Rate
+
+The Multichain protocol does not send two users in one transaction. All the transactions will be processed one by one. So there is only one transaction (from other chains to Cardano) in one block.
```
I'm not sure about the difference between the two documents. But the CIP style document mentions tx chaining.

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+## C. Bridging Cardano native token from Cardano to Ethereum
+
+1, User send native ADA to the SMPC address on Cardano and indicate the receiving address on Ethereum.
```
Since the user can just send any tokens/ada here. It'd be good to define how those tokens will be handled.

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+{
+    "metadata": {
+        "key": "123",
+        "value": {
+            "bind": "target address",
+            "toChainId": 1
+        }
```
How would the bridge know which token the user wants to bridge over? The user can include any number of tokens in the UTxO + ADA. The bridge would need to decide which one to take into consideration.

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
````diff
@@ -0,0 +1,395 @@
+```json
+{
+    "metadata": {
+        "key": "123",
````
Since the metadata is tied to a transaction, not an UTxO, there are additional requirements that need to be defined.
1. User includes multiple UTxO for the bridge address with different tokens in the same tx (e.g. they could send 2 different utxos, one containing 100 wUSDC and the second 1 wBTC)
2. user includes multiple UTxOs for the bridge address with the same token (2 utxos, 100 wUSDC and 250 wUSDC, in total 350 wUSDC)
3. user includes 1 UTxOs with 2 whitelisted assets: (100 wUSDC + 1 wBTC + for good measure 500 ADA)

TechRiderWR replied:
Response to a similar query by Luoxi. It'd be good to include this in the document:

> > 3. what if there are multiple tokens?
> > 4. what amount is considered here? Additional ada amount? what are the requirements?
> > 5. What’s is a legitimate range and who defines it?

> 3, Multiple tokens in one transaction is supported. Every token is treaded as a separate bridging transaction. txHash+index in the tokenList is one bridging transaction.
> 4, The amount of bridged token is considered here.
> 5, The legitimate range is between the minimum and maximum amount specified in the config contract, which is updated in the token deployment process.

TechRiderWR commented on CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md:
```diff
@@ -0,0 +1,395 @@
+
+# Flow Charts
+
+[How offchain cached utxo work.pdf](CIP%20Cardano%20Multichain%20Router%20b02eb0e3a2bf408a85bdbe6a4ce5603b/How_offchain_cached_utxo_work.pdf)
```
Additionally in the flow diagram. Do the MPC share a common utxo cache? What would happen if there are 2 swapins running at the same time and the MPC would sign 2 transactions using the same input UTxOs? Will thinks be retried?
